File: src/pyinterp/core/module/streaming_histogram.cpp
// Copyright (c) 2022 CNES
//
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
#include "pyinterp/streaming_histogram.hpp"

#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include <optional>

namespace py = pybind11;

template <typename Type>
void implement_streaming_histogram(py::module &m, const std::string &suffix) {
  py::class_<pyinterp::StreamingHistogram<Type>>(
      m, ("StreamingHistogram" + suffix).c_str(),
      ("StreamingHistogram" + suffix +
       "(self,"
       " values: numpy.ndarray[numpy.float32],"
       " weights: Optional[numpy.ndarray[numpy.float32]] = None,"
       " axis: Optional[List[int]] = None,"
       " bin_count: Optional[int] = None)" +
       R"__doc__(

Args:
    values: Array containing numbers whose statistics are desired.
    weights: An array of weights associated with the values. If not provided,
      all values are assumed to have equal weight.
    axis: Axis or axes along which to compute the statistics. If not provided,
      the statistics are computed over the flattened array.
    bin_count: Maximum number of bins to use to calculate the histogram.
      Default is 100.
)__doc__")
          .c_str())

      .def(py::init<py::array_t<Type, py::array::c_style> &,
                    std::optional<py::array_t<Type, py::array::c_style>> &,
                    std::optional<std::list<py::ssize_t>> &,
                    std::optional<size_t> &>(),
           py::arg("values"), py::arg("weights") = std::nullopt,
           py::arg("axis") = std::nullopt, py::arg("bin_count") = std::nullopt)
      .def("bins", &pyinterp::StreamingHistogram<Type>::bins,
           R"__doc__(
Returns the histogram bins.

Returns:
    The histogram bins.
)__doc__")
      .def("count", &pyinterp::StreamingHistogram<Type>::count,
           R"__doc__(
Returns the count of samples.

Returns:
    Count of samples.
      )__doc__")
      .def("kurtosis", &pyinterp::StreamingHistogram<Type>::kurtosis,
           R"__doc__(
Returns the kurtosis of samples.

Returns:
    Kurtosis of samples.
      )__doc__")
      .def("max", &pyinterp::StreamingHistogram<Type>::max,
           R"__doc__(
Returns maximum of samples.

Returns:
    Maximum of samples.
      )__doc__")
      .def("mean", &pyinterp::StreamingHistogram<Type>::mean,
           R"__doc__(
Returns the mean of samples.

Returns:
    Mean of samples.
      )__doc__")
      .def("min", &pyinterp::StreamingHistogram<Type>::min,
           R"__doc__(
Returns the minimum of samples.

Returns:
    Minimum of samples.
      )__doc__")
      .def("quantile", &pyinterp::StreamingHistogram<Type>::quantile,
           py::arg("q") = static_cast<Type>(0.5),
           R"__doc__(
Returns the quantile of samples.

Args:
    q: Quantile to compute.
Returns:
    Quantile of samples.
      )__doc__")
      .def("resize", &pyinterp::StreamingHistogram<Type>::resize,
           "Resize the maximum number of bins.")
      .def("size", &pyinterp::StreamingHistogram<Type>::size,
           R"__doc__(
Returns the number of bins.

Returns:
    Sum of bins.
      )__doc__")
      .def("sum_of_weights",
           &pyinterp::StreamingHistogram<Type>::sum_of_weights,
           R"__doc__(
Returns the sum of samples.

Returns:
    Sum of samples.
      )__doc__")
      .def("skewness", &pyinterp::StreamingHistogram<Type>::skewness,
           R"__doc__(
Returns the skewness of samples.

Returns:
    Skewness of samples.
      )__doc__")
      .def("variance", &pyinterp::StreamingHistogram<Type>::variance,
           R"__doc__(
Returns the variance of samples.

Returns:
    Variance of samples.
      )__doc__")
      .def(
          "__copy__",
          [](const pyinterp::StreamingHistogram<Type> &self) {
            return pyinterp::StreamingHistogram<Type>(self);
          },
          "Implements the shallow copy operation.",
          py::call_guard<py::gil_scoped_release>())
      .def("__iadd__", &pyinterp::StreamingHistogram<Type>::operator+=,
           py::arg("other"),
           "Overrides the default behavior of the ``+=`` operator.",
           py::call_guard<py::gil_scoped_release>())
      .def(
          "__add__",
          [](const pyinterp::StreamingHistogram<Type> &self,
             const pyinterp::StreamingHistogram<Type> &other) {
            auto result = pyinterp::StreamingHistogram<Type>(self);
            result += other;
            return result;
          },
          py::arg("other"),
          "Overrides the default behavior of the ``+`` operator.",
          py::call_guard<py::gil_scoped_release>())
      .def(py::pickle(
          [](const pyinterp::StreamingHistogram<Type> &self) {
            return self.getstate();
          },
          [](const py::tuple &state) {
            return pyinterp::StreamingHistogram<Type>::setstate(state);
          }));
}

void init_streaming_histogram(py::module &m) {
  implement_streaming_histogram<double>(m, "Float64");
  implement_streaming_histogram<float>(m, "Float32");
}

File: src/pyinterp/statistics/streaming_histogram.py
# Copyright (c) 2022 CNES
#
# All rights reserved. Use of this source code is governed by a
# BSD-style license that can be found in the LICENSE file.
"""
Calculate statistics of a stream of values
------------------------------------------
"""
from typing import Any, Iterable, Optional, Union

import dask.array.core
import numpy

from .. import core


def _delayed(
    attr: str,
    values: dask.array.core.Array,
    weights: Optional[dask.array.core.Array] = None,
    axis: Optional[Iterable[int]] = None,
    bin_count: Optional[int] = None,
) -> Union[core.StreamingHistogramFloat64, core.StreamingHistogramFloat32]:
    """Calculate the descriptive statistics of a dask array."""
    if weights is not None and values.shape != weights.shape:
        raise ValueError('values and weights must have the same shape')

    def _process_block(attr, x, w, axis, bin_count):
        instance = getattr(core, attr)(values=x,
                                       weights=w,
                                       axis=axis,
                                       bin_count=bin_count)
        return numpy.array([instance], dtype='object')

    drop_axis = list(range(values.ndim))[1:]

    return dask.array.core.map_blocks(
        _process_block,
        attr,
        values,
        weights,
        axis,
        bin_count,
        drop_axis=drop_axis,
        dtype='object').sum().compute()  # type: ignore


class StreamingHistogram:
    """Streaming histogram.

    The bins in the histogram have no predefined size, so that as values are
    pushed into the histogram, bins are added and merged as soon as their
    numbers exceed the maximum allowed capacity. A particularly interesting
    feature of streaming histograms is that they can be used to approximate
    quantiles without sorting (or even storing) values individually. The
    histograms can be constructed independently and merged, making them usable
    with Dask.

    Args:
        values: Array containing numbers whose statistics are desired.

            .. note::

                NaNs are automatically ignored.

        weights: An array of weights associated with the values. If not
            provided, all values are assumed to have equal weight.
        axis: Axis or axes along which to compute the statistics. If not
            provided, the statistics are computed over the flattened array.
        bin_count: The maximum number of bins to use in the histogram. If
            the number of bins exceeds the number of values, the histogram
            will be trimmed. Default is ``None``, which will set the number
            of bins to 100.
        dtype: Data type of the returned array. By default, the data type
            is numpy.float64.

    .. seealso::

        Yael Ben-Haim and Elad Tom-Tov,
        A Streaming Parallel Decision Tree Algorithm,
        Journal of Machine Learning Research, 11, 28, 849-872,
        http://jmlr.org/papers/v11/ben-haim10a.html

    .. note::

        If you do not want to estimate the quantiles of the dataset, use the
        class :py:class:`DescriptiveStatistics <pyinterp.DescriptiveStatistics>`
        which will give you more accurate results.
    """

    def __init__(self,
                 values: Union[dask.array.core.Array, numpy.ndarray],
                 weights: Optional[Union[dask.array.core.Array,
                                         numpy.ndarray]] = None,
                 axis: Optional[Union[int, Iterable[int]]] = None,
                 bin_count: Optional[int] = None,
                 dtype: Optional[numpy.dtype] = None) -> None:
        if isinstance(axis, int):
            axis = (axis, )
        dtype = dtype or numpy.dtype('float64')
        if dtype == numpy.dtype('float64'):
            attr = 'StreamingHistogramFloat64'
        elif dtype == numpy.dtype('float32'):
            attr = 'StreamingHistogramFloat32'
        else:
            raise ValueError(f'dtype {dtype} not handled by the object')
        if isinstance(values, dask.array.core.Array) or isinstance(
                weights, dask.array.core.Array):
            self._instance = _delayed(attr,
                                      dask.array.core.asarray(values),
                                      weights=dask.array.core.asarray(weights)
                                      if weights is not None else None,
                                      axis=axis,
                                      bin_count=bin_count)
        else:
            self._instance: Union[core.StreamingHistogramFloat64,
                                  core.StreamingHistogramFloat32] = getattr(
                                      core, attr)(values,
                                                  weights=weights,
                                                  axis=axis,
                                                  bin_count=bin_count)

    def __iadd__(self, other: Any) -> 'StreamingHistogram':
        """Adds a new histogram to the current one.

        Args:
            The histogram to add to the current one.

        Returns:
            itself.
        """
        if isinstance(other, StreamingHistogram):
            if type(self._instance) != type(other._instance):  # noqa: E721
                raise TypeError('StreamingHistogram types must match')
            self._instance += other._instance  # type: ignore
        else:
            raise TypeError('unsupported operand type(s) for +='
                            f": '{type(self)}' and '{type(other)}'")
        return self

    def bins(self) -> numpy.ndarray:
        """Returns the histogram bins.

        Returns:
            The histogram bins.
        """
        return self._instance.bins()

    def size(self) -> numpy.ndarray:
        """Returns the number of bins allocated to calculate the histogram.

        If :py:meth:`size() <pyinterp.StreamingHistogram.size>` is equal to
        :py:meth:`count() <pyinterp.StreamingHistogram.count>` then the
        histogram used to calculate the statistics is un-compressed. Otherwise,
        the histogram is compressed, which means that the calculated statistical
        quantities are an approximation of the statistical variables.

        Returns:
            Number of bins allocated to calculate the histogram.
        """
        return self._instance.size()

    def count(self) -> numpy.ndarray:
        """Returns the count of samples.

        Returns:
            The count of samples.
        """
        return self._instance.count()

    def kurtosis(self) -> numpy.ndarray:
        """Returns the kurtosis of samples.

        Returns:
            The kurtosis of samples.
        """
        return self._instance.kurtosis()

    def max(self) -> numpy.ndarray:
        """Returns the maximum of samples.

        Returns:
            The maximum of samples.
        """
        return self._instance.max()

    def mean(self) -> numpy.ndarray:
        """Returns the mean of samples.

        Returns:
            The mean of samples.
        """
        return self._instance.mean()

    def min(self) -> numpy.ndarray:
        """Returns the minimum of samples.

        Returns:
            The minimum of samples.
        """
        return self._instance.min()

    def skewness(self) -> numpy.ndarray:
        """Returns the skewness of samples.

        Returns:
            The skewness of samples.
        """
        return self._instance.skewness()

    def sum_of_weights(self) -> numpy.ndarray:
        """Returns the sum of weights.

        Returns:
            The sum of weights.
        """
        return self._instance.sum_of_weights()

    def var(self) -> numpy.ndarray:
        """Returns the variance of samples.

        Returns:
            The variance of samples.
        """
        return self._instance.variance()

    def std(self) -> numpy.ndarray:
        """Returns the standard deviation of samples.

        Returns:
            The standard deviation of samples.
        """
        return numpy.sqrt(self.var())

    def quantile(self, q: float = 0.5) -> numpy.ndarray:
        """Returns the q quantile of samples.

        Args:
            q (float): Quantile to compute. Default is ``0.5`` (median).

        Returns:
            The q quantile of samples.
        """
        return self._instance.quantile(q)
